src/GPUActivityAgent.cpp

```diff
@@ -338,14 +372,27 @@ namespace geopm
         if (!gpu_scoped_core_activity.empty()) {
             for (int domain_idx = 0; domain_idx < M_NUM_GPU; ++domain_idx) {
                 if (gpu_scoped_core_activity.at(domain_idx) >= M_GPU_ACTIVITY_CUTOFF) {
+                    // ROI proxy tracking
                     m_gpu_active_region_stop.at(domain_idx) = 0;
                     if (m_gpu_active_region_start.at(domain_idx) == 0) {
                         m_gpu_active_region_start.at(domain_idx) = m_time.value;
                         m_gpu_active_energy_start.at(domain_idx) = m_gpu_energy.at(domain_idx).value;
                     }
+
+                    // GPU on time tracking
+                    if (m_time.value > m_prev_time) {
+                        m_gpu_on_time.at(domain_idx) += m_time.value - m_prev_time;
+                    }
+
+                    // TODO: handle roll-over more gracefully than dropping a sample
+                    if (m_gpu_energy.at(domain_idx).value > m_prev_gpu_energy.at(domain_idx)) {
+                        m_gpu_on_energy.at(domain_idx) += m_gpu_energy.at(domain_idx).value - m_prev_gpu_energy.at(domain_idx);
+                    }
```

Comment:
I'm pretty surprised that this is necessary here and not handled by the IOGroup.
AFAICT, the GPU_ENERGY signal differs from the CPU_ENERGY signal because of this.  The MSRIOGroup automatically handles rollover of the other monotonic counters.  The GPU IOGroups also should do this.

The code in the agent could then be simplified, i.e. you don't need m_prev_gpu_energy at all if this was handled by the IOGroup.

Reply:
I agree this likely exists for CPU_ENERGY as well, but I have confirmed that the NVML and L0 IOGroups report the monotonic energy provided by the API (translated to SI units).  
We could add a "GPU_ENERGY_DELTA" signal (final naming TBD) using the geopm DifferenceSignal class that provides the difference in the last sample and current sample and handles the rollover.